tardis/utilities/executors/sshexecutor.py

```diff
@@ -25,18 +56,15 @@ async def run_command(self, command, stdin_input=None):
                 stdout=pe.stdout,
                 stderr=pe.stderr,
             ) from pe
-        except (
-            ConnectionResetError,
-            asyncssh.misc.DisconnectError,
-            asyncssh.misc.ConnectionLost,
-            BrokenPipeError,
-        ) as ce:
+        except asyncssh.ChannelOpenError as coe:
+            # Broken connection will be replaced by a new connection during next call
+            self._ssh_connection = None
```

Comment:
I am not sure if this can cause problems (a race condition) here? Please, review in particular.

Reply:
This depends at least on the fairness of ``self._lock``.

If ``self._lock`` is *not* fair, then a task A might be locked in ``initialize_connection``, while a task E holds the lock, creates the connection, and (assuming an unfair lock) releases the lock but uses the connection before A is woken up. This means that A assumes the connection exists, but the process of E may have failed and cleared the connection.

I recommend moving the management of the connection entirely into ``initialize_connection``, handling the connection via a getter instead of a side-effect:
* ``initialize_connection`` should return the current connection. ``run_command`` should unconditionally call ``initialize_connection`` and bind the connection to a local name. This avoids having to synchronise ``self._ssh_connection``.
* ``initialize_connection`` should use ``while`` instead of ``if`` to check the presence of the connection; it is fine to have a preceding, unlocked ``if`` as a fast-path if a connection does exist.

You might want to rename ``initialize_connection``, and make it a ``property`` similar to ``lock``.

Reply:
Thanks a lot for your suggestion. That was the solution I was looking for. :-)

Reply:
To make it a `property` unfortunately does not work, since it is a `coroutine`.